### ISSUE TITLE
Babel 7 + Typescript - extracting declaration `d.ts` file on transpile 

### ISSUE THREAD

**Issue by jaredvacanti** [babel/babel#9668]
## Feature Request

**Is your feature request related to a problem? Please describe.**
When bundling a library with webpack using some light typescript (using Babel and ESLint with their respective typescript integrations), I'd like to emit the type declarations file alongside the library.

`tsc` is not invoked in the workflow currently, and doing so would require another `tsconfig.json` due to conflicting settings associated with the `--emitDeclarationsOnly` flag. If [ESTree](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/typescript-estree) is handling the AST in order to remove the typescript and compile to es5, can the removed typescript be retained and extracted to a declarations file for later API consumption?

**Describe the solution you'd like**
An option to extract a declaration file 

**Describe alternatives you've considered**
Currently I write a standalone [declaration file](http://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html) (typically written to accompany projects written in javascript that are being incorporated into typescript projects).

**Comment by chrisblossom** [comment]
I would recommend creating an additional typescript config file [`tsconfig.types.json`](https://github.com/johnagan/clean-webpack-plugin/blob/2b94838bd140cbf9d6b6f5449de7d1e676ed87b2/tsconfig.types.json) that extends your base [`tsconfig.json`](https://github.com/johnagan/clean-webpack-plugin/blob/2b94838bd140cbf9d6b6f5449de7d1e676ed87b2/tsconfig.json) then running `tsc --project tsconfig.types.json` command during [your build process](https://github.com/johnagan/clean-webpack-plugin/blob/2b94838bd140cbf9d6b6f5449de7d1e676ed87b2/package.json#L31-L33).

**Comment by lll000111** [comment]
@chrisblossom As I understand it, and I'm pretty sure it is so, you can write whatever you want into your tsconfig.json used by `tsc` &mdash; because it is not read and used at all if you compile TypeScript via Babel. So you already have a tsconfig.json independent of whatever Babel+Typescript plugin does. Configure it to emit declaration files and simply run `tsc` to get them, independent of the .ts => .js transpilation done by Babel. I'm actually successfully experimenting with this very setup right now.

**Comment by chrisblossom** [comment]
> As I understand it, and I'm pretty sure it is so, you can write whatever you want into your tsconfig.json used by tsc — because it is not read and used at all if you compile TypeScript via Babel

This is correct as far as I know. The solution above does not use babel to emit declarations or use babel to do any type checking of any kind.

> Configure it to emit declaration files and simply run tsc to get them

When running `tsc` alone you wouldn't want to emit any files because that is what you use for type checking, well at least I do. Only need to emit declaration files during production build.

>https://devblogs.microsoft.com/typescript/typescript-and-babel-7/
> To make sure that TypeScript can call out some of these omissions, you should ensure that TypeScript uses the --isolatedModules flag.

The issue described in the post is because of conflicting settings, specifically with `isolatedModules: true`, and what you mentioned `allowJs: true`. Also wanting to ignore all `*.test.ts` files when emitting type declarations.



Looks like you are solving the problem the same way I did, just with CLI arguments instead of another configuration file, which is probably a nice way to unclutter the files. I'll probably look into this with my projects.

**Comment by nicolo-ribaudo** [comment]
This doesn't really fit Babel's scope.
Our goal is to be able to transpile TypeScript files to JavaScript so that you don't need to run both TS->JS and JS->old_JS during the same build task.

In order to generate `.d.ts` files, we would need to understand TypeScript's types semantics, which is currently not possible with our architecture. We recommend still using `tsc` to generate declaration files.

**Comment by EvertonSA** [comment]
I don't understand why are you guys recommending a lib that has a huge `This package has been deprecated`.

**Comment by nicolo-ribaudo** [comment]
The npm package is https://www.npmjs.com/package/typescript, which provides the `tsc` command.

**Comment by holt-mansfield-finaeo** [comment]
Babel transpiles my code in a few seconds. Then I run tsc to get types and wait over a minute.

I'm working in a monorepo and I have several watchers in place. I was hoping I could use babel to speed up my workflow. I need the typings because I'm using the code in other packages as I'm developing.

If babel could ever figure out how to emit types tsc would be a thing of the past.

**Comment by nicolo-ribaudo** [comment]
What makes TypeScript slower than Babel is not transforming your code: it's exactly figuring out types.
It's technically possible to create something around Babel (a plugin isn't enough, because it needs to analyze multiple files at the same time) which replaces `tsc`, but I think that you will hardly get any benefit for an enormous amount of added complexity.

**Comment by nicolo-ribaudo** [comment]
Closing per https://github.com/babel/babel/issues/9668#issuecomment-579464154.

TLDR: Babel can quickly compile your TypeScript code, but you should still use `tsc` in two cases:
- as a "linter" to ensure that your types are correct. You can run this before committing, while using Babel while developing.
- to generate .d.ts files.
